[16.0][ADD] stock_picking_batch_print_invoices

EmilioPascual opened the pull request:
This module was developed to be able to print invoices from batches and to be able print multiple copies.

It will be useful for you if you want to print all invoices included in a batch together.

It's neccesary https://github.com/OCA/stock-logistics-workflow/pull/1481 to merge.

@rafaelbn @Gelojr @Shide @yajo @fcvalgar please review, if you want 😉

@moduon MT-4407

EmilioPascual committed:
[ADD] stock_picking_Batch_print_invoices

@moduon MT-4407

Co-authored-by: Jairo Llopis <973709+yajo@users.noreply.github.com>

rousseldenis commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,32 @@
+                    _("The number of invoices copies must be greater than 0")
+                )
+
+    @api.onchange("batch_print_invoices")
```
Please prefer a stored field readonly=False than onchanges.

EmilioPascual replied:
OK, I changed this

rousseldenis commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,32 @@
+        for record in self:
+            if record.number_copies_invoices < 0:
+                raise ValidationError(
+                    _("The number of invoices copies must be greater than 0")
```
```suggestion
                    _("The number of invoices copies must be greater or equal to 0")
```

EmilioPascual replied:
You're right. Thank you.

yajo commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,38 @@
+
+    batch_print_invoices = fields.Boolean(
+        "Print Invoices from batch",
+        help="Allow print all invoices from batch",
```
**suggestion(dry):** this is redundant with the field name; either explain it better or remove the help:

```suggestion
```

yajo commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,38 @@
+    )
+    number_copies_invoices = fields.Integer(
+        "Number of invoices copies",
+        help="Number of invoices copies invoices to print",
```
**suggestion:** same as above, plus one typo:

```suggestion
        "Number of copies per invoice",
```

yajo commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,38 @@
+        for record in self:
+            if record.number_copies_invoices < 0:
+                raise ValidationError(
+                    _("The number of invoices copies must be greater or equal to 0")
```
```suggestion
                    _("The number of copies per invoice must be greater or equal to 0")
```

yajo commented on stock_picking_batch_print_invoices/models/stock_picking_type.py:
```diff
@@ -0,0 +1,38 @@
+            if record.batch_print_invoices:
+                record.number_copies_invoices = 1
+            else:
+                record.number_copies_invoices = 0
```
**suggestion:** remove this code and give `default=1` to `number_copies_invoices`.

You don't really care about the field value if invoices of this type aren't printed in batch, so you can make the field invisible in that case and just ignore its value. Otherwise, the constrain will make sure the value is valid.

yajo commented on stock_picking_batch_print_invoices/readme/CONTEXT.md:
```diff
@@ -0,0 +1,3 @@
+This module was developed to be able to print invoices from batches and to be able print multiple copies.
+
+It will be useful for you if you want to print all the delivery slips and invoices included in a batch together.
```
**suggestion:** give a better context and move this to description.

These paragraphs fit better in the description of the module, as they explain what the module does.

The context is [a different thing](https://gitlab.com/moduon/odoo-module-template/-/blob/66f9645417f59857874015367786324aa6596045/module/%7B%25%20if%20readme_fragments%20==%20'md'%20%25%7Dreadme%7B%25%20endif%20%25%7D/CONTEXT.md#L7). Example:

```suggestion
Food wholesalers have very little time to fill all their trucks with the goods that are planned to be sold on that day.

There is no time to waste when it comes to printing multiple delivery notes and invoices. When it is done, it is necessary to print multiple copies of these documents. It is faster to distribute them if they are all together.

This module was developed to cover this or a similar use case, and increase efficiency when filling many trucks.
```

yajo commented on stock_picking_batch_print_invoices/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1 @@
+This module extends the functionality of stock picking batch and to allow you to print multiple invoices from batches.
```
```suggestion
This module extends the functionality of stock batch pickings and to allows you to print multiple invoices from batches.
```

yajo commented on stock_picking_batch_print_invoices/readme/USAGE.md:
```diff
@@ -0,0 +1,19 @@
+1. Go to Inventory > Configuration > Operation Types.
+2. Choose one already created or create new operation type.
+3. Check print invoices.
+4. Fill in the number of copies you want.
```
**suggestion:** move to `CONFIGURE.md`.

Gelojr reviewed (changes requested):
Hi @EmilioPascual 
Nice job! But I think there is a behavior that is not correct. If there are several orders in an invoice that contain several delivery notes, the invoice is printed every time a delivery note is printed. And in the end, the invoice is printed as many times as there are delivery notes.
For example, if invoice F001 contains delivery notes A001, A002, and A003, it will be printed as follows:
A001
F001
A002
F001
A003
F001
https://www.loom.com/share/bd321c1fd6334327b3d660381b5e98ee?sid=46a70a72-fdc0-40c3-8500-346ad9366c4e
Can you check it please?

yajo commented on stock_picking_batch_print_invoices/tests/test_batch_print_invoices.py:
```diff
@@ -0,0 +1,197 @@
+    def test_stock_picking_batch_print_invoices_01(self):
+        """Batch without pickings, with any configuration in operation type,
+        batch mustn't print."""
+        for i in range(2):
+            for j in range(2):
+                for k in range(2):
+                    for m in range(2):
```
**suggestion:** use higher-order functions to reduce cyclomatic complexity:

```suggestion
        for i, j, k, m in itertools.product(*map(range, itertools.repeat(2, 4)))
```

Please, if possible, give variables more descriptive names.

yajo commented on stock_picking_batch_print_invoices/tests/test_batch_print_invoices.py:
```diff
@@ -0,0 +1,197 @@
+        self.batch.update(
+            {"picking_ids": [(6, 0, (self.picking_1 | self.picking_2).ids)]}
+        )
+        for i in range(2):
+            for j in range(2):
+                for k in range(2):
+                    for m in range(2):
```
**suggestion:** just like above:

```suggestion
        for i, j, k, m in itertools.product(*map(range, itertools.repeat(2, 4)))
```

yajo reviewed (approved):
A lot of suggestions but none of them is blocking. Good job!

EmilioPascual committed:
[FIX] stock_picking_group_by_partner_by_carrier: Added condition to check get_remaining_to_deliver

In some times, method get_remaining_to_deliver is not defined in template.

@moduon MT-4407

EmilioPascual commented:
Fix for stock_picking_group_by_partner_by_carrier to pass the tests: Sometimes method get_remaining_to_deliver is not defined, so it is checkedt first.

EmilioPascual committed:
[IMP] print pickings/invoices sorted by partner

@moduon MT-4427

EmilioPascual commented:
Improvement to print only once the invoice if several pickings are invoiced in single invoice.

yajo commented on stock_picking_batch_print_invoices/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1 @@
+This module extends the functionality of stock batch pickings and to allows you to print multiple invoices from batches.
```
```suggestion
This module extends the functionality of stock batch pickings to let you print multiple invoices from batches.
```

yajo reviewed (approved):
I just found one typo, but the rest looks very good.

rafaelbn reviewed (approved):
/ocabot merge patch
